File: src/workflow/modes/plan_act.py
# -*- coding: utf-8 -*-
"""
Plan-Act Mode - 计划-执行模式

实现结构化的计划-执行-审核工作流模式。
支持多阶段执行、检查点管理和迭代优化。
"""

import logging
import asyncio
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict, Any, Optional, Protocol, runtime_checkable
from datetime import datetime

logger = logging.getLogger("niko-workflow")


class WorkflowPhase(Enum):
    """工作流阶段"""
    PLAN = "plan"          # 计划阶段
    ACT = "act"            # 执行阶段
    REVIEW = "review"      # 审核阶段
    REVISE = "revise"      # 修订阶段
    COMPLETE = "complete"  # 完成


@dataclass
class PhaseResult:
    """阶段执行结果"""
    phase: WorkflowPhase
    success: bool
    output: Any = None
    metadata: Dict[str, Any] = field(default_factory=dict)
    error: Optional[Exception] = None
    duration_ms: int = 0
    next_phase: Optional[WorkflowPhase] = None

    @classmethod
    def ok(cls, phase: WorkflowPhase, output: Any, next_phase: Optional[WorkflowPhase] = None, **metadata) -> "PhaseResult":
        return cls(
            phase=phase,
            success=True,
            output=output,
            metadata=metadata,
            next_phase=next_phase
        )

    @classmethod
    def fail(cls, phase: WorkflowPhase, error: Exception) -> "PhaseResult":
        return cls(
            phase=phase,
            success=False,
            error=error
        )


@dataclass
class Checkpoint:
    """检查点"""
    phase: WorkflowPhase
    timestamp: datetime
    state: Dict[str, Any]
    iteration: int = 0

    def to_dict(self) -> Dict[str, Any]:
        return {
            "phase": self.phase.value,
            "timestamp": self.timestamp.isoformat(),
            "state": self.state,
            "iteration": self.iteration
        }


@dataclass
class PlanActState:
    """Plan-Act 模式状态"""
    session_id: str
    current_phase: WorkflowPhase = WorkflowPhase.PLAN
    iteration: int = 0
    max_iterations: int = 3
    plan: Optional[Dict[str, Any]] = None
    output: Optional[str] = None
    review_feedback: Optional[Dict[str, Any]] = None
    checkpoints: List[Checkpoint] = field(default_factory=list)
    phase_history: List[PhaseResult] = field(default_factory=list)

    def save_checkpoint(self) -> Checkpoint:
        """保存当前状态为检查点"""
        checkpoint = Checkpoint(
            phase=self.current_phase,
            timestamp=datetime.now(),
            state={
                "plan": self.plan,
                "output": self.output,
                "review_feedback": self.review_feedback,
            },
            iteration=self.iteration
        )
        self.checkpoints.append(checkpoint)
        return checkpoint

    def restore_checkpoint(self, checkpoint: Checkpoint) -> None:
        """恢复到检查点状态"""
        self.current_phase = checkpoint.phase
        self.iteration = checkpoint.iteration
        self.plan = checkpoint.state.get("plan")
        self.output = checkpoint.state.get("output")
        self.review_feedback = checkpoint.state.get("review_feedback")


@runtime_checkable
class IPhaseExecutor(Protocol):
    """阶段执行器协议"""

    async def execute(
        self,
        state: PlanActState,
        context: Dict[str, Any]
    ) -> PhaseResult:
        """执行阶段"""
        ...


class PlanPhaseExecutor:
    """
    计划阶段执行器

    使用 ArchitectAgent 生成执行计划。
    """

    def __init__(self, architect_agent: Optional[Any] = None):
        self._architect = architect_agent

    async def execute(
        self,
        state: PlanActState,
        context: Dict[str, Any]
    ) -> PhaseResult:
        """执行计划阶段"""
        start_time = datetime.now()

        try:
            task = context.get("task", "")
            requirements = context.get("requirements", [])

            # 如果有 ArchitectAgent，使用它生成计划
            if self._architect and hasattr(self._architect, 'plan'):
                plan = await self._architect.plan(
                    task=task,
                    requirements=requirements,
                    context=context
                )
            else:
                # 默认计划结构
                plan = {
                    "task": task,
                    "steps": [
                        {"step": 1, "action": "analyze_requirements", "description": "Analyze task requirements"},
                        {"step": 2, "action": "generate_content", "description": "Generate initial content"},
                        {"step": 3, "action": "refine_output", "description": "Refine and polish output"},
                    ],
                    "estimated_tokens": 2000,
                    "skills_required": context.get("skills", []),
                }

            state.plan = plan
            duration = int((datetime.now() - start_time).total_seconds() * 1000)

            return PhaseResult.ok(
                phase=WorkflowPhase.PLAN,
                output=plan,
                next_phase=WorkflowPhase.ACT,
                duration_ms=duration
            )

        except Exception as e:
            logger.error(f"Plan phase failed: {e}")
            return PhaseResult.fail(WorkflowPhase.PLAN, e)


class ActPhaseExecutor:
    """
    执行阶段执行器

    使用 WriterAgent 执行计划并生成内容。
    """

    def __init__(self, writer_agent: Optional[Any] = None):
        self._writer = writer_agent

    async def execute(
        self,
        state: PlanActState,
        context: Dict[str, Any]
    ) -> PhaseResult:
        """执行行动阶段"""
        start_time = datetime.now()

        try:
            if state.plan is None:
                raise ValueError("No plan available for execution")

            plan = state.plan
            task = context.get("task", "")

            # 如果有 WriterAgent，使用它生成内容
            if self._writer and hasattr(self._writer, 'write'):
                output = await self._writer.write(
                    task=task,
                    plan=plan,
                    context=context,
                    previous_feedback=state.review_feedback
                )
            else:
                # 默认输出
                steps_desc = "\n".join(f"- {s['description']}" for s in plan.get("steps", []))
                output = f"[Generated content for: {task}]\n\nExecuted steps:\n{steps_desc}"

            state.output = output
            duration = int((datetime.now() - start_time).total_seconds() * 1000)

            return PhaseResult.ok(
                phase=WorkflowPhase.ACT,
                output=output,
                next_phase=WorkflowPhase.REVIEW,
                duration_ms=duration
            )

        except Exception as e:
            logger.error(f"Act phase failed: {e}")
            return PhaseResult.fail(WorkflowPhase.ACT, e)


class ReviewPhaseExecutor:
    """
    审核阶段执行器

    使用 CriticAgent 评估输出质量。
    """

    def __init__(
        self,
        critic_agent: Optional[Any] = None,
        quality_threshold: float = 0.7
    ):
        self._critic = critic_agent
        self._quality_threshold = quality_threshold

    async def execute(
        self,
        state: PlanActState,
        context: Dict[str, Any]
    ) -> PhaseResult:
        """执行审核阶段"""
        start_time = datetime.now()

        try:
            if state.output is None:
                raise ValueError("No output to review")

            output = state.output

            # 如果有 CriticAgent，使用它评估
            if self._critic and hasattr(self._critic, 'evaluate'):
                feedback = await self._critic.evaluate(
                    content=output,
                    context=context
                )
            else:
                # 默认评估
                feedback = {
                    "overall_score": 0.75,
                    "dimensions": {
                        "coherence": 0.8,
                        "engagement": 0.7,
                        "style": 0.75,
                    },
                    "issues": [],
                    "suggestions": [],
                }

            state.review_feedback = feedback
            duration = int((datetime.now() - start_time).total_seconds() * 1000)

            # 决定下一阶段
            overall_score = feedback.get("overall_score", 0)
            if overall_score >= self._quality_threshold:
                next_phase = WorkflowPhase.COMPLETE
            elif state.iteration < state.max_iterations:
                next_phase = WorkflowPhase.REVISE
            else:
                next_phase = WorkflowPhase.COMPLETE
                logger.warning(f"Max iterations reached, accepting output with score {overall_score}")

            return PhaseResult.ok(
                phase=WorkflowPhase.REVIEW,
                output=feedback,
                next_phase=next_phase,
                duration_ms=duration,
                quality_score=overall_score
            )

        except Exception as e:
            logger.error(f"Review phase failed: {e}")
            return PhaseResult.fail(WorkflowPhase.REVIEW, e)


class RevisePhaseExecutor:
    """
    修订阶段执行器

    基于审核反馈修订输出。
    """

    def __init__(self, writer_agent: Optional[Any] = None):
        self._writer = writer_agent

    async def execute(
        self,
        state: PlanActState,
        context: Dict[str, Any]
    ) -> PhaseResult:
        """执行修订阶段"""
        start_time = datetime.now()

        try:
            if state.output is None or state.review_feedback is None:
                raise ValueError("Missing output or feedback for revision")

            # 如果有 WriterAgent，使用它修订
            if self._writer and hasattr(self._writer, 'revise'):
                revised = await self._writer.revise(
                    content=state.output,
                    feedback=state.review_feedback,
                    context=context
                )
            else:
                # 默认修订（简单返回原内容）
                issues = state.review_feedback.get("issues", [])
                revised = state.output
                if issues:
                    revised += f"\n\n[Revision notes: Addressed {len(issues)} issues]"

            state.output = revised
            state.iteration += 1
            duration = int((datetime.now() - start_time).total_seconds() * 1000)

            return PhaseResult.ok(
                phase=WorkflowPhase.REVISE,
                output=revised,
                next_phase=WorkflowPhase.REVIEW,
                duration_ms=duration,
                iteration=state.iteration
            )

        except Exception as e:
            logger.error(f"Revise phase failed: {e}")
            return PhaseResult.fail(WorkflowPhase.REVISE, e)


class PlanActMode:
    """
    Plan-Act 工作流模式

    实现结构化的计划-执行-审核循环。

    工作流程:
    1. Plan: 使用 ArchitectAgent 生成执行计划
    2. Act: 使用 WriterAgent 执行计划生成内容
    3. Review: 使用 CriticAgent 评估质量
    4. Revise (可选): 如果质量不达标，进行修订
    5. Complete: 返回最终输出

    使用示例:
        mode = PlanActMode(
            architect_agent=architect,
            writer_agent=writer,
            critic_agent=critic
        )

        result = await mode.execute(
            session_id="session-001",
            task="写一个悬疑开场",
            context={"genre": "mystery"}
        )

        if result.success:
            print(result.output)
    """

    def __init__(
        self,
        architect_agent: Optional[Any] = None,
        writer_agent: Optional[Any] = None,
        critic_agent: Optional[Any] = None,
        quality_threshold: float = 0.7,
        max_iterations: int = 3
    ):
        self._plan_executor = PlanPhaseExecutor(architect_agent)
        self._act_executor = ActPhaseExecutor(writer_agent)
        self._review_executor = ReviewPhaseExecutor(critic_agent, quality_threshold)
        self._revise_executor = RevisePhaseExecutor(writer_agent)

        self._executors: Dict[WorkflowPhase, IPhaseExecutor] = {
            WorkflowPhase.PLAN: self._plan_executor,
            WorkflowPhase.ACT: self._act_executor,
            WorkflowPhase.REVIEW: self._review_executor,
            WorkflowPhase.REVISE: self._revise_executor,
        }

        self._max_iterations = max_iterations
        self._states: Dict[str, PlanActState] = {}

    async def execute(
        self,
        session_id: str,
        task: str,
        context: Optional[Dict[str, Any]] = None,
        resume_from_checkpoint: bool = False
    ) -> PhaseResult:
        """
        执行 Plan-Act 工作流

        Args:
            session_id: 会话 ID
            task: 任务描述
            context: 额外上下文
            resume_from_checkpoint: 是否从检查点恢复

        Returns:
            最终阶段结果
        """
        # 获取或创建状态
        if session_id in self._states and resume_from_checkpoint:
            state = self._states[session_id]
            if state.checkpoints:
                state.restore_checkpoint(state.checkpoints[-1])
        else:
            state = PlanActState(
                session_id=session_id,
                max_iterations=self._max_iterations
            )
            self._states[session_id] = state

        # 合并上下文
        full_context = {"task": task}
        if context:
            full_context.update(context)

        # 执行工作流循环
        while state.current_phase != WorkflowPhase.COMPLETE:
            executor = self._executors.get(state.current_phase)

            if executor is None:
                logger.error(f"No executor for phase: {state.current_phase}")
                break

            # 保存检查点
            state.save_checkpoint()

            # 执行阶段
            result = await executor.execute(state, full_context)
            state.phase_history.append(result)

            if not result.success:
                logger.error(f"Phase {state.current_phase.value} failed: {result.error}")
                return result

            # 转移到下一阶段
            if result.next_phase:
                state.current_phase = result.next_phase
            else:
                break

        # 返回最终结果
        return PhaseResult.ok(
            phase=WorkflowPhase.COMPLETE,
            output=state.output,
            plan=state.plan,
            review_feedback=state.review_feedback,
            iterations=state.iteration,
            phase_count=len(state.phase_history)
        )

    def get_state(self, session_id: str) -> Optional[PlanActState]:
        """获取会话状态"""
        return self._states.get(session_id)

    def clear_state(self, session_id: str) -> bool:
        """清除会话状态"""
        if session_id in self._states:
            del self._states[session_id]
            return True
        return False

    def list_sessions(self) -> List[str]:
        """列出所有会话"""
        return list(self._states.keys())


# ============ 便捷函数 ============

def get_default_plan_act_mode(
    architect_agent: Optional[Any] = None,
    writer_agent: Optional[Any] = None,
    critic_agent: Optional[Any] = None
) -> PlanActMode:
    """获取预配置的 Plan-Act 模式"""
    return PlanActMode(
        architect_agent=architect_agent,
        writer_agent=writer_agent,
        critic_agent=critic_agent,
        quality_threshold=0.7,
        max_iterations=3
    )

File: src-ts/workflow/modes/plan-act.ts
/**
 * Plan-Act Mode - Structured Plan-Execute-Review workflow pattern.
 *
 * Supports multi-phase execution, checkpoint management, and iterative refinement.
 *
 * Migrated from src/workflow/modes/plan_act.py
 */

import { createLogger } from '../../logger/index.js';

const _log = createLogger('workflow-plan-act');

// ============================================================
// WorkflowPhase enum
// ============================================================

export enum WorkflowPhase {
  PLAN = 'plan',
  ACT = 'act',
  REVIEW = 'review',
  REVISE = 'revise',
  COMPLETE = 'complete',
}

// ============================================================
// PhaseResult
// ============================================================

export interface PhaseResult {
  phase: WorkflowPhase;
  success: boolean;
  output: unknown;
  metadata: Record<string, unknown>;
  error: Error | null;
  durationMs: number;
  nextPhase: WorkflowPhase | null;
}

function okPhaseResult(
  phase: WorkflowPhase,
  output: unknown,
  nextPhase: WorkflowPhase | null = null,
  metadata?: Record<string, unknown>,
  durationMs = 0,
): PhaseResult {
  return {
    phase,
    success: true,
    output,
    metadata: metadata ?? {},
    error: null,
    durationMs,
    nextPhase,
  };
}

function failPhaseResult(
  phase: WorkflowPhase,
  error: Error,
): PhaseResult {
  return {
    phase,
    success: false,
    output: null,
    metadata: {},
    error,
    durationMs: 0,
    nextPhase: null,
  };
}

// ============================================================
// Checkpoint
// ============================================================

export interface Checkpoint {
  phase: WorkflowPhase;
  timestamp: string;
  state: Record<string, unknown>;
  iteration: number;
}

function checkpointToDict(cp: Checkpoint): Record<string, unknown> {
  return {
    phase: cp.phase,
    timestamp: cp.timestamp,
    state: cp.state,
    iteration: cp.iteration,
  };
}

// ============================================================
// PlanActState
// ============================================================

export class PlanActState {
  sessionId: string;
  currentPhase: WorkflowPhase;
  iteration: number;
  maxIterations: number;
  plan: Record<string, unknown> | null;
  output: string | null;
  reviewFeedback: Record<string, unknown> | null;
  checkpoints: Checkpoint[];
  phaseHistory: PhaseResult[];

  constructor(sessionId: string, maxIterations = 3) {
    this.sessionId = sessionId;
    this.currentPhase = WorkflowPhase.PLAN;
    this.iteration = 0;
    this.maxIterations = maxIterations;
    this.plan = null;
    this.output = null;
    this.reviewFeedback = null;
    this.checkpoints = [];
    this.phaseHistory = [];
  }

  /** Save current state as a checkpoint. */
  saveCheckpoint(): Checkpoint {
    const checkpoint: Checkpoint = {
      phase: this.currentPhase,
      timestamp: new Date().toISOString(),
      state: {
        plan: this.plan,
        output: this.output,
        review_feedback: this.reviewFeedback,
      },
      iteration: this.iteration,
    };
    this.checkpoints.push(checkpoint);
    return checkpoint;
  }

  /** Restore state from a checkpoint. */
  restoreCheckpoint(checkpoint: Checkpoint): void {
    this.currentPhase = checkpoint.phase;
    this.iteration = checkpoint.iteration;
    this.plan = (checkpoint.state['plan'] as Record<string, unknown>) ?? null;
    this.output = (checkpoint.state['output'] as string) ?? null;
    this.reviewFeedback = (checkpoint.state['review_feedback'] as Record<string, unknown>) ?? null;
  }
}

// ============================================================
// IPhaseExecutor interface (replaces Python Protocol)
// ============================================================

export interface IPhaseExecutor {
  execute(state: PlanActState, context: Record<string, unknown>): Promise<PhaseResult>;
}

// ============================================================
// Agent interfaces (duck-typed, matching Python hasattr checks)
// ============================================================

export interface IArchitectAgent {
  plan?(task: string, requirements: unknown[], context: Record<string, unknown>): Promise<Record<string, unknown>>;
}

export interface IWriterAgent {
  write?(task: string, plan: Record<string, unknown>, context: Record<string, unknown>, previousFeedback: Record<string, unknown> | null): Promise<string>;
  revise?(content: string, feedback: Record<string, unknown>, context: Record<string, unknown>): Promise<string>;
}

export interface ICriticAgent {
  evaluate?(content: string, context: Record<string, unknown>): Promise<Record<string, unknown>>;
}

// ============================================================
// PlanPhaseExecutor
// ============================================================

export class PlanPhaseExecutor implements IPhaseExecutor {
  private _architect: IArchitectAgent | null;

  constructor(architectAgent?: IArchitectAgent | null) {
    this._architect = architectAgent ?? null;
  }

  async execute(state: PlanActState, context: Record<string, unknown>): Promise<PhaseResult> {
    const startTime = Date.now();

    try {
      const task = String(context['task'] ?? '');
      const requirements = context['requirements'] as unknown[] ?? [];

      let plan: Record<string, unknown>;

      if (this._architect && typeof this._architect.plan === 'function') {
        plan = await this._architect.plan(task, requirements, context);
      } else {
        // Default plan structure
        plan = {
          task,
          steps: [
            { step: 1, action: 'analyze_requirements', description: 'Analyze task requirements' },
            { step: 2, action: 'generate_content', description: 'Generate initial content' },
            { step: 3, action: 'refine_output', description: 'Refine and polish output' },
          ],
          estimated_tokens: 2000,
          skills_required: context['skills'] ?? [],
        };
      }

      state.plan = plan;
      const duration = Date.now() - startTime;

      return okPhaseResult(
        WorkflowPhase.PLAN,
        plan,
        WorkflowPhase.ACT,
        { duration_ms: duration },
        duration,
      );
    } catch (e) {
      const error = e instanceof Error ? e : new Error(String(e));
      _log.error(`Plan phase failed: ${error.message}`);
      return failPhaseResult(WorkflowPhase.PLAN, error);
    }
  }
}

// ============================================================
// ActPhaseExecutor
// ============================================================

export class ActPhaseExecutor implements IPhaseExecutor {
  private _writer: IWriterAgent | null;

  constructor(writerAgent?: IWriterAgent | null) {
    this._writer = writerAgent ?? null;
  }

  async execute(state: PlanActState, context: Record<string, unknown>): Promise<PhaseResult> {
    const startTime = Date.now();

    try {
      if (state.plan === null) {
        throw new Error('No plan available for execution');
      }

      const plan = state.plan;
      const task = String(context['task'] ?? '');

      let output: string;

      if (this._writer && typeof this._writer.write === 'function') {
        output = await this._writer.write(task, plan, context, state.reviewFeedback);
      } else {
        const steps = plan['steps'] as Array<Record<string, unknown>> ?? [];
        const stepsDesc = steps.map(s => `- ${s['description']}`).join('\n');
        output = `[Generated content for: ${task}]\n\nExecuted steps:\n${stepsDesc}`;
      }

      state.output = output;
      const duration = Date.now() - startTime;

      return okPhaseResult(
        WorkflowPhase.ACT,
        output,
        WorkflowPhase.REVIEW,
        { duration_ms: duration },
        duration,
      );
    } catch (e) {
      const error = e instanceof Error ? e : new Error(String(e));
      _log.error(`Act phase failed: ${error.message}`);
      return failPhaseResult(WorkflowPhase.ACT, error);
    }
  }
}

// ============================================================
// ReviewPhaseExecutor
// ============================================================

export class ReviewPhaseExecutor implements IPhaseExecutor {
  private _critic: ICriticAgent | null;
  private _qualityThreshold: number;

  constructor(criticAgent?: ICriticAgent | null, qualityThreshold = 0.7) {
    this._critic = criticAgent ?? null;
    this._qualityThreshold = qualityThreshold;
  }

  async execute(state: PlanActState, context: Record<string, unknown>): Promise<PhaseResult> {
    const startTime = Date.now();

    try {
      if (state.output === null) {
        throw new Error('No output to review');
      }

      let feedback: Record<string, unknown>;

      if (this._critic && typeof this._critic.evaluate === 'function') {
        feedback = await this._critic.evaluate(state.output, context);
      } else {
        feedback = {
          overall_score: 0.75,
          dimensions: {
            coherence: 0.8,
            engagement: 0.7,
            style: 0.75,
          },
          issues: [],
          suggestions: [],
        };
      }

      state.reviewFeedback = feedback;
      const duration = Date.now() - startTime;

      // Determine next phase
      const overallScore = typeof feedback['overall_score'] === 'number'
        ? feedback['overall_score'] as number
        : 0;
      let nextPhase: WorkflowPhase;
      if (overallScore >= this._qualityThreshold) {
        nextPhase = WorkflowPhase.COMPLETE;
      } else if (state.iteration < state.maxIterations) {
        nextPhase = WorkflowPhase.REVISE;
      } else {
        nextPhase = WorkflowPhase.COMPLETE;
        _log.warn(`Max iterations reached, accepting output with score ${overallScore}`);
      }

      return okPhaseResult(
        WorkflowPhase.REVIEW,
        feedback,
        nextPhase,
        { duration_ms: duration, quality_score: overallScore },
        duration,
      );
    } catch (e) {
      const error = e instanceof Error ? e : new Error(String(e));
      _log.error(`Review phase failed: ${error.message}`);
      return failPhaseResult(WorkflowPhase.REVIEW, error);
    }
  }
}

// ============================================================
// RevisePhaseExecutor
// ============================================================

export class RevisePhaseExecutor implements IPhaseExecutor {
  private _writer: IWriterAgent | null;

  constructor(writerAgent?: IWriterAgent | null) {
    this._writer = writerAgent ?? null;
  }

  async execute(state: PlanActState, context: Record<string, unknown>): Promise<PhaseResult> {
    const startTime = Date.now();

    try {
      if (state.output === null || state.reviewFeedback === null) {
        throw new Error('Missing output or feedback for revision');
      }

      let revised: string;

      if (this._writer && typeof this._writer.revise === 'function') {
        revised = await this._writer.revise(state.output, state.reviewFeedback, context);
      } else {
        const issues = state.reviewFeedback['issues'] as unknown[] ?? [];
        revised = state.output;
        if (issues.length > 0) {
          revised += `\n\n[Revision notes: Addressed ${issues.length} issues]`;
        }
      }

      state.output = revised;
      state.iteration += 1;
      const duration = Date.now() - startTime;

      return okPhaseResult(
        WorkflowPhase.REVISE,
        revised,
        WorkflowPhase.REVIEW,
        { duration_ms: duration, iteration: state.iteration },
        duration,
      );
    } catch (e) {
      const error = e instanceof Error ? e : new Error(String(e));
      _log.error(`Revise phase failed: ${error.message}`);
      return failPhaseResult(WorkflowPhase.REVISE, error);
    }
  }
}

// ============================================================
// PlanActMode - Orchestrator
// ============================================================

export class PlanActMode {
  private _planExecutor: PlanPhaseExecutor;
  private _actExecutor: ActPhaseExecutor;
  private _reviewExecutor: ReviewPhaseExecutor;
  private _reviseExecutor: RevisePhaseExecutor;

  private _executors: Map<WorkflowPhase, IPhaseExecutor>;
  private _maxIterations: number;
  private _states: Map<string, PlanActState>;

  constructor(
    options?: {
      architectAgent?: IArchitectAgent;
      writerAgent?: IWriterAgent;
      criticAgent?: ICriticAgent;
      qualityThreshold?: number;
      maxIterations?: number;
    },
  ) {
    this._planExecutor = new PlanPhaseExecutor(options?.architectAgent);
    this._actExecutor = new ActPhaseExecutor(options?.writerAgent);
    this._reviewExecutor = new ReviewPhaseExecutor(options?.criticAgent, options?.qualityThreshold ?? 0.7);
    this._reviseExecutor = new RevisePhaseExecutor(options?.writerAgent);

    this._executors = new Map<WorkflowPhase, IPhaseExecutor>([
      [WorkflowPhase.PLAN, this._planExecutor],
      [WorkflowPhase.ACT, this._actExecutor],
      [WorkflowPhase.REVIEW, this._reviewExecutor],
      [WorkflowPhase.REVISE, this._reviseExecutor],
    ]);

    this._maxIterations = options?.maxIterations ?? 3;
    this._states = new Map();
  }

  /**
   * Execute the Plan-Act workflow.
   *
   * Flow:
   * 1. Plan: ArchitectAgent generates execution plan
   * 2. Act: WriterAgent executes plan and generates content
   * 3. Review: CriticAgent evaluates quality
   * 4. Revise (optional): If quality is insufficient, revise
   * 5. Complete: Return final output
   */
  async execute(
    sessionId: string,
    task: string,
    context?: Record<string, unknown>,
    resumeFromCheckpoint = false,
  ): Promise<PhaseResult> {
    // Get or create state
    let state: PlanActState;
    if (this._states.has(sessionId) && resumeFromCheckpoint) {
      state = this._states.get(sessionId)!;
      if (state.checkpoints.length > 0) {
        state.restoreCheckpoint(state.checkpoints[state.checkpoints.length - 1]);
      }
    } else {
      state = new PlanActState(sessionId, this._maxIterations);
      this._states.set(sessionId, state);
    }

    // Merge context
    const fullContext: Record<string, unknown> = { task };
    if (context) {
      Object.assign(fullContext, context);
    }

    // Execute workflow loop
    while (state.currentPhase !== WorkflowPhase.COMPLETE) {
      const executor = this._executors.get(state.currentPhase);

      if (!executor) {
        _log.error(`No executor for phase: ${state.currentPhase}`);
        break;
      }

      // Save checkpoint
      state.saveCheckpoint();

      // Execute phase
      const result = await executor.execute(state, fullContext);
      state.phaseHistory.push(result);

      if (!result.success) {
        _log.error(`Phase ${state.currentPhase} failed: ${result.error?.message}`);
        return result;
      }

      // Transition to next phase
      if (result.nextPhase != null) {
        state.currentPhase = result.nextPhase;
      } else {
        break;
      }
    }

    // Return final result
    return okPhaseResult(
      WorkflowPhase.COMPLETE,
      state.output,
      null,
      {
        plan: state.plan,
        review_feedback: state.reviewFeedback,
        iterations: state.iteration,
        phase_count: state.phaseHistory.length,
      },
    );
  }

  /** Get the state for a session. */
  getState(sessionId: string): PlanActState | undefined {
    return this._states.get(sessionId);
  }

  /** Clear a session's state. */
  clearState(sessionId: string): boolean {
    return this._states.delete(sessionId);
  }

  /** List all session IDs. */
  listSessions(): string[] {
    return Array.from(this._states.keys());
  }
}

// ============================================================
// Convenience factory
// ============================================================

export function getDefaultPlanActMode(
  architectAgent?: IArchitectAgent,
  writerAgent?: IWriterAgent,
  criticAgent?: ICriticAgent,
): PlanActMode {
  return new PlanActMode({
    architectAgent,
    writerAgent,
    criticAgent,
    qualityThreshold: 0.7,
    maxIterations: 3,
  });
}
